### PR TITLE
optparse: switch to zlist_t

### PR DESCRIPTION
**Pull request by grondo**
(split off from #2894)

optparse is the only lib in flux-core still using the liblsd/list implementation.

While the liblsd/list interface is arguably better than zlist/x, it has some drawbacks like the static internal allocation pool that can't freed at exit (not a big deal, but slightly more confusing for valgrind tests), and an internal mutex. It is also a bit strange to have a secondary list implementation for this one purpose.

This PR converts liboptparse to zlist/zlistx internally, and drops list.[ch] from the source.

### CODE REVIEW

**Comment by grondo** [comment]
I accidentally pushed my branch to flux-framework/flux-core at first, so now the `Travis-CI - Branch` check is failing (since the branch is deleted). This can be ignored, but sorry about that!

File: src/common/liboptparse/optparse.c [hunk 1]
```diff
@@ -983,9 +1000,12 @@ optparse_err_t optparse_add_option_table (optparse_t *p,
     optparse_err_t rc = OPTPARSE_SUCCESS;
     const struct optparse_option *o = opts;
 
+    if (!p)
+        return OPTPARSE_BAD_ARG;
+
     while (o->usage) {
         if ((rc = optparse_add_option (p, o++)) != OPTPARSE_SUCCESS) {
-            option_table_remove (p, opts, o);
+            option_table_remove (p, opts, --o);
```
Comment (chu11) [inline src/common/liboptparse/optparse.c:1008]: i see the bug you caught here.  But should it be in a different commit?

Comment (grondo) [reply src/common/liboptparse/optparse.c:1008]: Good catch. Yeah I forgot I fixed that along the way. I'll split this out.

Comment (grondo) [reply src/common/liboptparse/optparse.c]: Ok, latest push should have that fix split out into its own commit.

Comment (chu11) [reply src/common/liboptparse/optparse.c]: LGTM, but did you mean "Fix the value for o (letter o) passed to ..."?  You did "0" (zero) instead.

**Comment by garlick** [comment]
Is there a reason not to use `zlistx_t` for both the option list and the option arguments?

**Comment by grondo** [comment]
> Is there a reason not to use zlistx_t for both the option list and the option arguments?

zlistx is a bit of a pain to use, so I only use it when I need a find function. (fyi there was already a use of `zlist_t` in optparse.c for some reason.)

Is there some reason zlistx_t is preferred to zlist_t for simple lists? If so I can convert the whole file over to that.

**Comment by garlick** [comment]
We're using both throughout `flux-core` and I usually go for `zlist_t` in simple cases too.   It just struck me as odd to use both in the same source module.  Given that this PR is just getting rid of `List`, feel free to ignore.

**Comment by grondo** [comment]
Not the end of the world to do a linear search with `zlist_t` if it would be preferable...

**Comment by garlick** [comment]
> Not the end of the world to do a linear search with zlist_t if it would be preferable...

Slightly (to me) but  I don't think it's a big deal.

**Comment by grondo** [comment]
Eh, I think I agree with you. Plus, I was thinking it might be nice if options were sorted in order of insertion into the option table, not in alphanumeric order, so a zlistx_t wouldn't make sense in that case.

File: src/common/liboptparse/optparse.c (hunk 1 above)
Comment (grondo) [reply src/common/liboptparse/optparse.c]: Thanks, fixed.

**Comment by grondo** [comment]
Ok, I switched the `zlistx_t` use over to `zlist_t`, squashed and force-pushed the result.

**Comment by grondo** [comment]
Thanks!
